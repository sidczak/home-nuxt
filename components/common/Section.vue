<template lang="pug">
section.d-flex.flex-shrink-0.align-items-center.section-py-80(:class="[{'section-min-h-100': minHeight, 'animated': scrolled}, typeAnimation]")
    slot(v-if="noBody" name="content")
    b-container(v-else)
        b-row.justify-content-center.mb-2
            b-col.text-center(md="6")
                h2 {{title}}
                h3 {{subtitle}}
                p.mb-0(v-if="desc" v-html="desc")
        slot(name="content")
</template>
<script>
export default {
    props: {
        noBody: {
            type: Boolean,
            default: false,
        },
        onAnimation: {
            type: Boolean,
            default: false,
        },
        typeAnimation: {
            type: String,
            default: null,
            validator: (value) =>
                ["blur", "tada"].includes(value) || null === value,
        },
        title: {
            type: String,
            default: null,
        },
        subtitle: {
            type: String,
            default: null,
        },
        desc: {
            type: String,
            default: null,
        },
        minHeight: {
            type: Boolean,
            default: true,
        },
    },
    data() {
        return {
            scrolled: false,
            // sectionPosition: 0,
        };
    },
    mounted() {
        window.addEventListener("scroll", this.handleScroll);
        this.handleScroll(); // ustawi scrolled=true dla pierwszej sekcji.
    },
    destroyed() {
        window.removeEventListener("scroll", this.handleScroll);
    },
    methods: {
        handleScroll() {
            const section = this.$el;
            // const scrollStart = section.offsetTop - 200;
            // const scrollStart = section.offsetTop - (window.innerHeight * 0.75); // 25% wczytanej sekcji
            // const scrollStart = section.offsetTop - (window.innerHeight * 50/100); // 50% wczytanej sekcji
            const scrollStart = section.offsetTop - window.innerHeight / 2; // 50% wczytajej sekcji
            // this.sectionPosition = window.scrollY - section.offsetTop;

            if (this.onAnimation && scrollStart <= window.scrollY) {
                this.scrolled = true;
            } else {
                this.scrolled = false;
            }
        },
    },
};
</script>
<style lang="scss" scoped>
@import "~sass-rem";
$rem-fallback: false;
section {
    position: relative;
    // @include rem(padding, 6px 30px);
    // font-size: rem(11px);
    &.tada {
        opacity: 0;
        transition: all 0.5s;
        &.animated {
            opacity: 1;
        }
    }
    &.blur {
        opacity: 0;
        filter: blur(5px);
        transition: all 0.5s;
        &.animated {
            opacity: 1;
            filter: blur(0);
        }
    }
}
.section {
    &-pt-152 {
        padding-top: rem(152px);
    }
    &-py-80,
    &-pt-80 {
        padding-top: rem(80px);
    }
    &-py-80,
    &-pb-80 {
        padding-bottom: rem(80px);
    }
    &-min-h-100 {
        min-height: 100%;
    }
}
</style>
