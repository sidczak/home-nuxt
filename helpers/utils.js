const kebabToCamelCase = (s) => s.replace(/-./g, (x) => x[1].toUpperCase());

const camelCaseToKebab = (s) =>
    s
        .replace(/((?<=[a-z\d])[A-Z]|(?<=[A-Z\d])[A-Z](?=[a-z]))/g, "-$1")
        .toLowerCase();

const hasValue = (value) => value !== undefined && null !== value;

const getVh = (v) => {
    const h = Math.max(
        document.documentElement.clientHeight,
        window.innerHeight || 0
    );
    return (v * h) / 100;
};

const getCountryCode = (code) => {
    const dict = {
        _1: "XK", //_1 - Kosowo
    };
    return dict[code] || code;
};

const extractFileName = (path) => {
    const parts = path.split("/");
    return parts[parts.length - 1].split(".")[0];
};

const importAsyncComponents = (componentsContext) => {
    const components = {};
    componentsContext.keys().forEach((path) => {
        components[extractFileName(path)] = () => componentsContext(path);
    });
    return components;
};

const registerDirectives = (Vue, context) => {
    context.keys().forEach((path) => {
        const camelCaseName = camelCaseToKebab(extractFileName(path));
        Vue.directive(camelCaseName, context(path).default);
    });
};

const getTranslations = (Vue, translationsKeysDict) => {
    const translations = {};
    Object.keys(translationsKeysDict).forEach((key) => {
        translations[key] = Vue.$t(translationsKeysDict[key]);
    });
    return translations;
};

const omitObjectKeys = (obj, keys) => {
    const newObj = {};
    Object.keys(obj).forEach((key) => {
        if (!keys.includes(key)) {
            newObj[key] = obj[key];
        }
    });
    return newObj;
};

const omitObjectKeysArray = (arr, keys) => {
    return arr.map((object) => {
        return omitObjectKeys(object, keys);
    });
};

const changeAllDictValues = (object, value) => {
    const preparedObject = {};
    Object.keys(object).forEach((key) => {
        preparedObject[key] = value;
    });
    return preparedObject;
};

const getLastItem = (arr) => {
    return arr[arr.length - 1];
};

const hexToRgb = (hex) =>
    hex
        .replace(
            /^#?([a-f\d])([a-f\d])([a-f\d])$/i,
            (m, r, g, b) => "#" + r + r + g + g + b + b
        )
        .substring(1)
        .match(/.{2}/g)
        .map((x) => parseInt(x, 16));

const getResponseData = (response, dataKeyHeaderName = "x-data-key") => {
    const dataKey = response.headers[dataKeyHeaderName];
    if (!dataKey) {
        throw new Error(
            `"${dataKeyHeaderName}" header is not present in response headers`
        );
    }
    if (!(dataKey in response.data)) {
        throw new Error(
            `"${dataKeyHeaderName}" key is not present in response data`
        );
    }
    return response.data[dataKey];
};

const roundNumber = (value, decimalPlaces = 0) => {
    const p = Math.pow(10, decimalPlaces);
    return Math.round((value + Number.EPSILON) * p) / p;
};

const range = (length, fromZero = false) => {
    return Array.from({ length }, (x, i) => (fromZero ? i : i + 1));
};

const isFunction = (value) => {
    return "function" === typeof value;
};

module.exports = {
    kebabToCamelCase,
    camelCaseToKebab,
    hasValue,
    getVh,
    getCountryCode,
    extractFileName,
    importAsyncComponents,
    registerDirectives,
    getTranslations,
    omitObjectKeys,
    omitObjectKeysArray,
    changeAllDictValues,
    getLastItem,
    hexToRgb,
    getResponseData,
    roundNumber,
    isFunction,
    range,
};
