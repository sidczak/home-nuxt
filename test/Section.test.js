import Section from "../components/common/Section.vue";
import { mount } from "@vue/test-utils";
import { awaitNTimes, getTree } from "./utils";

import Vue from "vue";
import BootstrapVue from "bootstrap-vue";
Vue.use(BootstrapVue);

describe("Section", () => {
    it("It completed props", async () => {
        const wrapper = mount(Section, {
            propsData: {
                title: "title",
                subtitle: "subtitle",
            },
        });
        const tree = getTree(wrapper.html());
        expect(tree).toMatchSnapshot();
    });

    it("It completed props2", async () => {
        const wrapper = mount(Section, {
            propsData: {
                title: "title2",
                subtitle: "subtitle2",
            },
        });
        const tree = getTree(wrapper.html());
        expect(tree).toMatchSnapshot();
    });
});
