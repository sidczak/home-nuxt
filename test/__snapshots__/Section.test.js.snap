// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section It completed props 1`] = `
{
  "$el": "<section class="d-flex flex-shrink-0 align-items-center section-py-80 section-min-h-100">
  <div class="container">
    <div class="row justify-content-center mb-2">
      <div class="text-center col-md-6">
        <h2>title</h2>
        <h3>subtitle</h3>
        <!---->
      </div>
    </div>
  </div>
</section>",
}
`;

exports[`Section It completed props2 1`] = `
{
  "$el": "<section class="d-flex flex-shrink-0 align-items-center section-py-80 section-min-h-100">
  <div class="container">
    <div class="row justify-content-center mb-2">
      <div class="text-center col-md-6">
        <h2>title2</h2>
        <h3>subtitle2</h3>
        <!---->
      </div>
    </div>
  </div>
</section>",
}
`;
